PumaGrasshopper/AttributeForm.cs

```diff
@@ -12,25 +12,158 @@ namespace PumaGrasshopper
+                    {
+                        defaultValueMap.GetBool(attributeName, out bool val);
+                        return val.ToString();
+                    }
+                case Annotations.AnnotationArgumentType.AAT_FLOAT:
+                    {
+                        defaultValueMap.GetDouble(attributeName, out double val);
+                        return val.ToString();
+                    }
+                case Annotations.AnnotationArgumentType.AAT_STR:
+                    {
+                        defaultValueMap.GetString(attributeName, out string val);
+                        return val;
+                    }
+                case Annotations.AnnotationArgumentType.AAT_INT:
+                    {
+                        defaultValueMap.GetInteger(attributeName, out int val);
+                        return val.ToString();
+                    }
+                case Annotations.AnnotationArgumentType.AAT_BOOL_ARRAY:
+                    {
+                        defaultValueMap.GetBoolArray(attributeName, out bool[] vals);
+                        return vals.Aggregate("", (left, right) => left + right + ";");
+                    }
+                case Annotations.AnnotationArgumentType.AAT_FLOAT_ARRAY:
+                    {
+                        defaultValueMap.GetDoubleArray(attributeName, out double[] vals);
+                        return vals.Aggregate("", (left, right) => left + right + ";");
+                    }
+                case Annotations.AnnotationArgumentType.AAT_STR_ARRAY:
+                    {
+                        defaultValueMap.GetStringArray(attributeName, out string[] vals);
+                        return vals.Aggregate("", (left, right) => left + right + ";");
+                    }
+                case Annotations.AnnotationArgumentType.AAT_INT_ARRAY:
+                    {
+                        defaultValueMap.GetIntegerArray(attributeName, out int[] vals);
+                        return vals.Aggregate("", (left, right) => left + right + ";");
+                    }
+                default:
+                    return "No default value";
+            }
+        }
+
+        private List<ListViewItem> getAttributesListViewItems(List<RuleAttribute> ruleAttributes)
+        {
+            List<ListViewItem> listViewItems = ruleAttributes.ConvertAll(
+                attribute => {
+                    return new ListViewItem(new string[] {
+                            attribute.mNickname,
+                            GetTypeString(attribute.mAttribType),
+                            GetDefaultValueString(attribute.mFullName, attribute.mAttribType)
+                        },
+                        group: Groups.Find(group => group.Header == attribute.mGroup));
+                    }
+                );
+
+            return listViewItems;
+        }
+
+        private ListViewGroup[] getAllGroups(List<RuleAttribute> attributes)

```

Comment:
I think you could return a List<ListViewGroup> here because you are converting it to a List at the callsite again anyways.